### PR TITLE
Plt metadata url

### PR DESCRIPTION
**Pull request by soerenbf**
## Purpose

Please be critical of non-idiomatic implementations.

## Changes

-  Implements `TokenMetadataUrl` along with its CBOR representation

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: haskell-tests/Types/CBOR.hs [hunk 1]
```diff
@@ -295,17 +313,101 @@ testEncodedTokenOperations = describe "EncodedTokenOperations JSON serialization
...
+                  TokenMetadataUrl
+                    { tmUrl = "https://abc.token/meta",
+                      tmChecksumSha256 = Just $ SHA256.Hash (FBS.pack $ replicate 32 0xab),
+                      tmAdditional = Map.empty
+                    }
+                )
+            )
+            ( deserialiseFromBytes
+                decodeTokenMetadataUrl
+                "\xA2\x63\x75\x72\x6C\x76\x68\x74\x74\x70\x73\x3A\x2F\x2F\x61\x62\x63\x2E\x74\x6F\x6B\x65\x6E\x2F\x6D\x65\x74\x61\x6E\x63\x68\x65\x63\x6B\x73\x75\x6D\x53\x68\x61\x32\x35\x36\x58\x20\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB\xAB"
```
Comment (soerenbf) [inline haskell-tests/Types/CBOR.hs:353]: I got weird mismatches when I tried to represent this as a string as done in the other tests. 

`"\x6EchecksumSha256"` would get the hex representation `"ec6865636b73756d5368613235365"` which is incorrect (should be `"6e636865636b73756d5368613235365`".

To avoid this I use the raw byte representation instead.

This goes for the other tests added as well.

File: haskell-tests/Types/CBOR.hs [hunk 2]
```diff
@@ -153,26 +171,26 @@ testInitializationParameters = describe "token-initialization-parameters decodin
...
                 "\xA4\x64name\x69\
-                \ABC token\x68metadata\x76https://abc.token/meta\x69\
+                \ABC token\x68metadata\xA1\x63url\x76https://abc.token/meta\x69\
                 \allowList\xF4\x6DinitialSupply\xC4\x82\x24\x19\x27\x10"
             )
     it "Missing \"name\"" $
         assertEqual
             "Decoded CBOR"
-            (Left (DeserialiseFailure 64 "Missing \"name\""))
+            (Left (DeserialiseFailure 69 "Missing \"name\""))
```
Comment (soerenbf) [inline haskell-tests/Types/CBOR.hs:180]: Byte offset changes by 5 bytes due to the changes to the 5 bytes added `"\xA1\x63url"` to the cbor representation of the metadata url

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 3]
```diff
@@ -211,6 +216,113 @@ encodeSequence encodeItem s =
...
+hexToCborTerm :: Text -> Either String CBOR.Term
+hexToCborTerm hexText = do
+    bs <- Base16.decode (TextEncoding.encodeUtf8 hexText)
+    decodeTerm bs
+  where
+    decodeTerm bs =
+        case CBOR.deserialiseFromBytes CBOR.decodeTerm (LBS.fromStrict bs) of
+            Left err -> Left $ "Failed to decode CBOR term: " ++ show err
+            Right ("", term) -> Right term
+            Right (remaining, _) -> Left $ "Extra bytes after decoding CBOR term: " ++ show (LBS.length remaining)
```
Comment (soerenbf) [inline haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:245]: Could move these helper functions to local scopes instead. I left them here as they might be generally applicable in future composite types, but if we don't want that I can move them.

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 4]
```diff
@@ -1280,9 +1416,18 @@ decodeTokenModuleState = decodeMap decodeVal build Map.empty
...
     convertText _ = Nothing
+
     convertBool (CBOR.TBool b) = Just b
     convertBool _ = Nothing
 
+    -- Convert CBOR to TokenMetadataUrl
+    convertTokenMetadataUrl :: CBOR.Term -> Maybe TokenMetadataUrl
+    convertTokenMetadataUrl term =
+        -- TODO: Would be nice not to have to re-encode the term to bytes...
+        case CBOR.deserialiseFromBytes decodeTokenMetadataUrl (CBOR.toLazyByteString $ CBOR.encodeTerm term) of
```
Comment (td202) [inline haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:1427]:
```suggestion
        case decodeTokenMetadataUrlHelper term of
```

File: haskell-tests/Types/CBOR.hs [hunk 5]
```diff
@@ -6,35 +6,47 @@ import Codec.CBOR.Read
...
-
 genText :: Gen Text.Text
 genText = sized $ \s -> Text.decodeUtf8 . BS.pack <$> genUtf8String s
 
 genTokenAmount :: Gen TokenAmount
 genTokenAmount = TokenAmount <$> arbitrary <*> chooseBoundedIntegral (0, 255)
 
+genSha256Hash :: Gen SHA256.Hash
+genSha256Hash = do
+    randomBytes <- BS.pack <$> replicateM SHA256.digestSize (choose (0, 255)) -- Generate 32 random bytes
```
Comment (td202) [inline haskell-tests/Types/CBOR.hs:37]:
```suggestion
    randomBytes <- BS.pack <$> vector SHA256.digestSize -- Generate 32 random bytes
```

File: haskell-tests/Types/CBOR.hs [hunk 6]
```diff
@@ -6,35 +6,47 @@ import Codec.CBOR.Read
...
+    return (SHA256.hash randomBytes)
+
+genTokenMetadataUrl :: Gen TokenMetadataUrl
+genTokenMetadataUrl = do
+    url <- genText
+    checksumSha256 <- Just <$> genSha256Hash
+    return TokenMetadataUrl{tmUrl = url, tmChecksumSha256 = checksumSha256, tmAdditional = Map.empty}
```
Comment (td202) [inline haskell-tests/Types/CBOR.hs:44]: I think we also need tests for the case where `tmAdditional` is non-empty (cf. `genTGokenModuleStateWithAdditional`).

File: haskell-tests/Types/CBOR.hs (hunk 1 above)
Comment (td202) [reply haskell-tests/Types/CBOR.hs:353]: I also recommend splitting this across a few lines.

File: haskell-tests/Types/CBOR.hs [hunk 7]
```diff
@@ -295,17 +313,101 @@ testEncodedTokenOperations = describe "EncodedTokenOperations JSON serialization
...
             === ( deserialiseFromBytes
                     decodeTokenTransfer
                     (toLazyByteString $ encodeTokenTransfer tt)
                 )
+    it "Encode and decode TokenMetadataUrl" $ withMaxSuccess 1000 $ forAll genTokenMetadataUrl $ \tt ->
+        (Right ("", tt))
+            === ( deserialiseFromBytes
+                    decodeTokenMetadataUrl
+                    (toLazyByteString $ encodeTokenMetadataUrl tt)
+                )
```
Comment (td202) [inline haskell-tests/Types/CBOR.hs:410]: I recommend also doing this with JSON encoding, and having tests with/without additional fields. (Could make the generator 50/50 on whether additional fields are included, rather than having two separate generators.)

**Review by td202 — APPROVED** [review]

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 8]
```diff
@@ -211,6 +216,137 @@ encodeSequence encodeItem s =
...
         <> foldMap encodeItem s
 
+convertSha256Hash :: CBOR.Term -> Maybe SHA256.Hash
+convertSha256Hash (CBOR.TBytes bs)
+    | BS.length bs == SHA256.digestSize = Just $ SHA256.Hash (FBS.fromByteString bs)
+    | otherwise = Nothing
+convertSha256Hash _ = Nothing
+
+encodeSha256Hash :: SHA256.Hash -> Encoding
+encodeSha256Hash (SHA256.Hash h) =
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:226]: Moved this to local scope

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 9]
```diff
@@ -211,6 +216,137 @@ encodeSequence encodeItem s =
     encodeListLen (fromIntegral $ Seq.length s)
         <> foldMap encodeItem s
 
+convertSha256Hash :: CBOR.Term -> Maybe SHA256.Hash
+convertSha256Hash (CBOR.TBytes bs)
+    | BS.length bs == SHA256.digestSize = Just $ SHA256.Hash (FBS.fromByteString bs)
+    | otherwise = Nothing
+convertSha256Hash _ = Nothing
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:223]: Moved it to local scope

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 10]
```diff
@@ -211,6 +216,137 @@ encodeSequence encodeItem s =
...
+decodeTokenMetadataUrl = decodeMap decodeVal build Map.empty
+  where
+    decodeVal key = Just $ mapValueDecoder key CBOR.decodeTerm (at key)
+    build :: Map.Map Text CBOR.Term -> Either String TokenMetadataUrl
+    build m0 = do
+        (tmUrl, m1) <- getAndClear "url" convertText m0
+        (tmChecksumSha256, tmAdditional) <- getMaybeAndClear "checksumSha256" convertSha256Hash m1
+        return TokenMetadataUrl{..}
+    getAndClear key convert m = do
+        let (maybeTerm, m') = m & at key <<.~ Nothing
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:307]: Ideally this would be documented on the infix operator itself would it not? This is more or less copied from another place in the code so I assumed that it was fine to leave undocumented. I agree that it's confusing though, but I'm also confused by pretty much all infix operators used in the code base, so that's not saying much 😅

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs [hunk 11]
```diff
@@ -211,6 +216,137 @@ encodeSequence encodeItem s =
...
+createTokenMetadataUrlWithSha256 :: Text -> SHA256.Hash -> TokenMetadataUrl
+createTokenMetadataUrlWithSha256 url checksum = TokenMetadataUrl{tmUrl = url, tmChecksumSha256 = Just checksum, tmAdditional = Map.empty}
+
+instance AE.ToJSON TokenMetadataUrl where
+    toJSON TokenMetadataUrl{..} =
+        AE.object $
+            [ "url" AE..= tmUrl,
+              "checksumSha256" AE..= tmChecksumSha256
+            ]
+                ++ map (\(k, v) -> AE.Key.fromText k AE..= cborTermToHex v) (Map.toList tmAdditional)
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:274]: Is a `Nothing` not omitted by default? I might misunderstand this, but in the source code of Aeson it defines the following implementation of `ToJSON` for `Maybe a`:

```
instance ToJSON1 Maybe where
    liftToJSON _ t _ (Just a) = t a
    liftToJSON _ _  _ Nothing  = Null

    liftToEncoding _ t _ (Just a) = t a
    liftToEncoding _ _  _ Nothing  = E.null_

    liftOmitField _ = isNothing

instance (ToJSON a) => ToJSON (Maybe a) where
    toJSON = toJSON1
    omitField = omitField1
    toEncoding = toEncoding1
```

Does this not mean that if `Nothing` is found, the field will be omitted, as per `omitField` delegating to `isNothing`?

File: haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs:307]: I added a comment explaining `m & at key <<.~ Nothing`
